From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 1 Jan 2022 19:56:02 +1000
Subject: [PATCH] Mark permissions as dirty when modified


diff --git a/src/main/java/org/bukkit/permissions/PermissibleBase.java b/src/main/java/org/bukkit/permissions/PermissibleBase.java
index 30b98d1645c571ba5c18e5cc93b0bec3f74b1d3b..34bc958a5c5bcddb4e5cbe992e1639f8988ed11f 100644
--- a/src/main/java/org/bukkit/permissions/PermissibleBase.java
+++ b/src/main/java/org/bukkit/permissions/PermissibleBase.java
@@ -21,6 +21,7 @@ public class PermissibleBase implements Permissible {
     private final Permissible parent;
     private final List<PermissionAttachment> attachments = new LinkedList<PermissionAttachment>();
     private final Map<String, PermissionAttachmentInfo> permissions = new HashMap<String, PermissionAttachmentInfo>();
+    public boolean dirty = false; // MultiPaper
 
     public PermissibleBase(@Nullable ServerOperator opable) {
         this.opable = opable;
@@ -177,6 +178,8 @@ public class PermissibleBase implements Permissible {
         for (PermissionAttachment attachment : attachments) {
             calculateChildPermissions(attachment.getPermissions(), false, attachment);
         }
+
+        dirty = true; // MultiPaper
     }
 
     public synchronized void clearPermissions() {
