From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 8 Mar 2022 21:43:27 +1000
Subject: [PATCH] Store the known entity


diff --git a/src/main/java/net/minecraft/world/level/entity/PersistentEntitySectionManager.java b/src/main/java/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
index 6baa313b8201ed23193d7885c85606b0899ade3c..6c09dca06eaae1d6afbf1bc022a6f632400722f9 100644
--- a/src/main/java/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
+++ b/src/main/java/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
@@ -14,11 +14,7 @@ import it.unimi.dsi.fastutil.objects.ObjectIterator;
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.io.Writer;
-import java.util.List;
-import java.util.Objects;
-import java.util.Queue;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
 import java.util.concurrent.CompletableFuture;
 import java.util.function.Consumer;
 import java.util.stream.Collectors;
@@ -40,7 +36,7 @@ import org.bukkit.event.entity.EntityRemoveEvent;
 public class PersistentEntitySectionManager<T extends EntityAccess> implements AutoCloseable {
 
     static final Logger LOGGER = LogUtils.getLogger();
-    final Set<UUID> knownUuids = Sets.newHashSet();
+    public final HashMap<UUID, T> knownUuids = new HashMap<>(); // MultiPaper - store the known entity
     final LevelCallback<T> callbacks;
     public final EntityPersistentStorage<T> permanentStorage;
     private final EntityLookup<T> visibleEntityStorage = new EntityLookup<>();
@@ -79,7 +75,7 @@ public class PersistentEntitySectionManager<T extends EntityAccess> implements A
 
     private boolean addEntityUuid(T entity) {
         org.spigotmc.AsyncCatcher.catchOp("Entity add by UUID"); // Paper
-        if (!this.knownUuids.add(entity.getUUID())) {
+        if (this.knownUuids.put(entity.getUUID(), entity) != null) {
             PersistentEntitySectionManager.LOGGER.warn("UUID of added entity already exists: {}", entity);
             return false;
         } else {
@@ -389,7 +385,7 @@ public class PersistentEntitySectionManager<T extends EntityAccess> implements A
     }
 
     public boolean isLoaded(UUID uuid) {
-        return this.knownUuids.contains(uuid);
+        return this.knownUuids.containsKey(uuid);
     }
 
     public LevelEntityGetter<T> getEntityGetter() {
