From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: neonjava <neonjavadev@gmail.com>
Date: Sun, 12 Jul 2026 16:03:11 +0530
Subject: [PATCH] Sync level.dat


diff --git a/src/main/java/net/minecraft/world/level/storage/LevelStorageSource.java b/src/main/java/net/minecraft/world/level/storage/LevelStorageSource.java
index 85ba843ce7e1f62971e736fa2cc028c47b274ce4..785ccdbca65f9479a58f00aa58cff1317155cd2b 100644
--- a/src/main/java/net/minecraft/world/level/storage/LevelStorageSource.java
+++ b/src/main/java/net/minecraft/world/level/storage/LevelStorageSource.java
@@ -271,11 +271,20 @@ public class LevelStorageSource {
     }
 
     static CompoundTag readLevelDataTagRaw(Path path) throws IOException {
-        return NbtIo.readCompressed(path, NbtAccounter.create(104857600L));
+        // MultiPaper start
+        CompoundTag nbttagcompound = puregero.multipaper.MultiPaper.readLevel(path.toFile().getParentFile().getName());
+        if (nbttagcompound == null) {
+            LOGGER.info("A new level.dat is being generated at " + path.toFile().getPath() + ", forcing it to save");
+            puregero.multipaper.MultiPaper.levelDatNeedsSaving = true;
+            return null;
+        }
+        return nbttagcompound;
+        // MultiPaper end
     }
 
     static Dynamic<?> readLevelDataTagFixed(Path path, DataFixer dataFixer) throws IOException {
         CompoundTag nbttagcompound = LevelStorageSource.readLevelDataTagRaw(path);
+        if (nbttagcompound == null) return null; // MultiPaper
         CompoundTag nbttagcompound1 = nbttagcompound.getCompound("Data");
         int i = NbtUtils.getDataVersion(nbttagcompound1, -1); final int version = i; // Paper - obfuscation helpers
         Dynamic<?> dynamic = DataFixTypes.LEVEL.updateToCurrentVersion(dataFixer, new Dynamic(NbtOps.INSTANCE, nbttagcompound1), i);
@@ -380,6 +389,7 @@ public class LevelStorageSource {
 
     @Nullable
     private static Tag readLightweightData(Path path) throws IOException {
+        if (true) return puregero.multipaper.MultiPaper.readLevel(path.toFile().getParentFile().getName()); // MultiPaper
         SkipFields skipfields = new SkipFields(new FieldSelector[]{new FieldSelector("Data", CompoundTag.TYPE, "Player"), new FieldSelector("Data", CompoundTag.TYPE, "WorldGenSettings")});
 
         NbtIo.parseCompressed(path, skipfields, NbtAccounter.create(104857600L));
@@ -612,6 +622,12 @@ public class LevelStorageSource {
             Path path = this.levelDirectory.path();
 
             try {
+                // MultiPaper start
+                if (true) {
+                    puregero.multipaper.MultiPaper.writeLevel(path.toFile().getName(), nbt);
+                    return;
+                }
+                // MultiPaper end
                 Path path1 = Files.createTempFile(path, "level", ".dat");
 
                 NbtIo.writeCompressed(nbt, path1);
@@ -753,7 +769,11 @@ public class LevelStorageSource {
         }
 
         public boolean hasWorldData() {
-            return Files.exists(this.levelDirectory.dataFile(), new LinkOption[0]) || Files.exists(this.levelDirectory.oldDataFile(), new LinkOption[0]);
+            try {
+                return puregero.multipaper.MultiPaper.readLevel(this.levelId) != null;
+            } catch (IOException e) {
+                return false;
+            }
         }
 
         public void close() throws IOException {
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index bc92ede80abbb1f72fbc6f14caa9d9cc4cfff7a6..779c7c8a349c0e916abe83a5657981b4101c31ed 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -28,9 +28,18 @@ import java.util.concurrent.TimeUnit;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
 
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import net.minecraft.server.level.ServerLevel;
+import puregero.multipaper.mastermessagingprotocol.messages.masterbound.ReadLevelMessage;
+import puregero.multipaper.mastermessagingprotocol.messages.masterbound.WriteLevelMessage;
+import puregero.multipaper.mastermessagingprotocol.messages.serverbound.DataMessageReply;
+
 public class MultiPaper {
 
     private static MultiPaperConnection multiPaperConnection = null;
+    private static final Logger LOGGER = LogManager.getLogger(MultiPaper.class.getSimpleName());
+    public static boolean levelDatNeedsSaving = false;
 
     public static MultiPaperConnection getConnection() {
         if (multiPaperConnection == null) {
@@ -58,6 +67,14 @@ public class MultiPaper {
                 server.getConnection().tick();
             }
         }
+
+        if (levelDatNeedsSaving) {
+            levelDatNeedsSaving = false;
+            LOGGER.info("A level.dat needs saving, all worlds are being force saved");
+            for (ServerLevel level : MinecraftServer.getServer().getAllLevels()) {
+                level.saveIncrementally(true);
+            }
+        }
     }
 
     public static void sendTickTime(long time, double tps) {
@@ -165,6 +182,18 @@ public class MultiPaper {
         return isExternalPlayer(((CraftEntity) bukkitEntity).getHandle());
     }
 
+    public static CompoundTag readLevel(String world) throws IOException {
+        byte[] data = getConnection().sendAndAwaitReply(new ReadLevelMessage(world), DataMessageReply.class).thenApply(message -> message.data).join();
+
+        return data.length == 0 ? null : NbtIo.readCompressed(new ByteArrayInputStream(data));
+    }
+
+    public static void writeLevel(String world, CompoundTag compoundTag) throws IOException {
+        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+        NbtIo.writeCompressed(compoundTag, buffer);
+        getConnection().send(new WriteLevelMessage(world, buffer.toByteArray()), message -> { /* do nothing */ });
+    }
+
     public static byte[] nbtToBytes(CompoundTag compoundTag) throws IOException {
         if (compoundTag == null) {
             return new byte[0];
