From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: neonjava <neonjavadev@gmail.com>
Date: Sun, 12 Jul 2026 16:07:55 +0530
Subject: [PATCH] Sync JSON files and restore missing chunk syncing files


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java b/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
index 22c4f8dea99f92a1eb3da2baf0a15bf9d2ca0462..f8571e86d02b34baa620aa71bd21aba86cbf99da 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
@@ -23,16 +23,18 @@ public class DedicatedPlayerList extends PlayerList {
     @Override
     public void loadAndSaveFiles() {
         // Paper end - fix converting txt to json file
+        // MultiPaper start - why are we saving these if we haven't modified them?
         this.loadUserBanList();
-        this.saveUserBanList();
+        // this.saveUserBanList();
         this.loadIpBanList();
-        this.saveIpBanList();
+        // this.saveIpBanList();
         this.loadOps();
         this.loadWhiteList();
-        this.saveOps();
-        if (!this.getWhiteList().getFile().exists()) {
-            this.saveWhiteList();
-        }
+        // this.saveOps();
+        // if (!this.getWhiteList().getFile().exists()) {
+        //     this.saveWhiteList();
+        // }
+        // MultiPaper end
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/server/players/IpBanList.java b/src/main/java/net/minecraft/server/players/IpBanList.java
index 8382b5de75803e2d5443662cedc975e548c36fbc..c65f6ebea2507011b732c64c2518f531e17c3c7f 100644
--- a/src/main/java/net/minecraft/server/players/IpBanList.java
+++ b/src/main/java/net/minecraft/server/players/IpBanList.java
@@ -5,6 +5,9 @@ import java.io.File;
 import java.net.SocketAddress;
 import javax.annotation.Nullable;
 
+import puregero.multipaper.externalserverprotocol.BanIpPacket;
+import puregero.multipaper.externalserverprotocol.PardonIpPacket;
+
 public class IpBanList extends StoredUserList<String, IpBanListEntry> {
     public IpBanList(File file) {
         super(file);
@@ -42,4 +45,18 @@ public class IpBanList extends StoredUserList<String, IpBanListEntry> {
 
         return string;
     }
+
+    // MultiPaper start
+    @Override
+    public void add(IpBanListEntry entry) {
+        super.add(entry);
+        BanIpPacket.broadcast(entry.getUser(), entry.created, entry.expires, entry.reason, entry.source);
+    }
+
+    @Override
+    public void remove(String key) {
+        super.remove(key);
+        PardonIpPacket.broadcast(key);
+    }
+    // MultiPaper end
 }
diff --git a/src/main/java/net/minecraft/server/players/ServerOpList.java b/src/main/java/net/minecraft/server/players/ServerOpList.java
index a1c9686043b5a8c5cb1614b46e10484000c920ae..934a5f63d70709c10e990b2246ffc7e0d40525f8 100644
--- a/src/main/java/net/minecraft/server/players/ServerOpList.java
+++ b/src/main/java/net/minecraft/server/players/ServerOpList.java
@@ -5,6 +5,9 @@ import com.mojang.authlib.GameProfile;
 import java.io.File;
 import java.util.Objects;
 
+import puregero.multipaper.externalserverprotocol.DeopPlayerPacket;
+import puregero.multipaper.externalserverprotocol.OpPlayerPacket;
+
 public class ServerOpList extends StoredUserList<GameProfile, ServerOpListEntry> {
     public ServerOpList(File file) {
         super(file);
@@ -29,4 +32,18 @@ public class ServerOpList extends StoredUserList<GameProfile, ServerOpListEntry>
     protected String getKeyForUser(GameProfile gameProfile) {
         return gameProfile.getId().toString();
     }
+
+    // MultiPaper start
+    @Override
+    public void add(ServerOpListEntry entry) {
+        super.add(entry);
+        OpPlayerPacket.broadcast(entry.getUser().getName(), entry.getUser().getId(), entry.getLevel(), entry.getBypassesPlayerLimit());
+    }
+
+    @Override
+    public void remove(GameProfile key) {
+        super.remove(key);
+        DeopPlayerPacket.broadcast(key.getName(), key.getId());
+    }
+    // MultiPaper end
 }
diff --git a/src/main/java/net/minecraft/server/players/StoredUserList.java b/src/main/java/net/minecraft/server/players/StoredUserList.java
index c038da20b76c0b7b1c18471b20be01e849d29f3a..7ac5326014835f8175f80c9ba060e08b0c4e27a1 100644
--- a/src/main/java/net/minecraft/server/players/StoredUserList.java
+++ b/src/main/java/net/minecraft/server/players/StoredUserList.java
@@ -25,6 +25,8 @@ import net.minecraft.Util;
 import net.minecraft.util.GsonHelper;
 import org.slf4j.Logger;
 
+import puregero.multipaper.MultiPaper;
+
 public abstract class StoredUserList<K, V extends StoredUserEntry<K>> {
 
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -114,10 +116,10 @@ public abstract class StoredUserList<K, V extends StoredUserEntry<K>> {
 
         Objects.requireNonNull(jsonarray);
         stream.forEach(jsonarray::add);
-        BufferedWriter bufferedwriter = Files.newWriter(this.file, StandardCharsets.UTF_8);
+        BufferedWriter bufferedwriter = null; // Files.newWriter(this.file, StandardCharsets.UTF_8); // MultiPaper - don't touch files
 
         try {
-            StoredUserList.GSON.toJson(jsonarray, StoredUserList.GSON.newJsonWriter(bufferedwriter));
+            puregero.multipaper.MultiPaper.writeJson(this.file.getName(), StoredUserList.GSON.toJson(jsonarray)); // MultiPaper
         } catch (Throwable throwable) {
             if (bufferedwriter != null) {
                 try {
@@ -137,14 +139,21 @@ public abstract class StoredUserList<K, V extends StoredUserEntry<K>> {
     }
 
     public void load() throws IOException {
-        if (this.file.exists()) {
-            BufferedReader bufferedreader = Files.newReader(this.file, StandardCharsets.UTF_8);
+        if (true || this.file.exists()) { // MultiPaper
+            BufferedReader bufferedreader = null; // Files.newReader(this.file, StandardCharsets.UTF_8); // MultiPaper
+            
+            // MultiPaper start
+            String json = MultiPaper.readJson(this.file.getName());
+            if (json == null || json.isEmpty()) {
+                return;
+            }
+            // MultiPaper end
 
             label54:
             {
                 try {
                     this.map.clear();
-                    JsonArray jsonarray = (JsonArray) StoredUserList.GSON.fromJson(bufferedreader, JsonArray.class);
+                    JsonArray jsonarray = (JsonArray) StoredUserList.GSON.fromJson(json, JsonArray.class); // MultiPaper
 
                     if (jsonarray == null) {
                         break label54;
diff --git a/src/main/java/net/minecraft/server/players/UserBanList.java b/src/main/java/net/minecraft/server/players/UserBanList.java
index 7ece29a64fc44cf5d5ec4ec22a084fdd7b9554bf..cf887a580923464af755e6e00b07d607efe020c4 100644
--- a/src/main/java/net/minecraft/server/players/UserBanList.java
+++ b/src/main/java/net/minecraft/server/players/UserBanList.java
@@ -5,6 +5,9 @@ import com.mojang.authlib.GameProfile;
 import java.io.File;
 import java.util.Objects;
 
+import puregero.multipaper.externalserverprotocol.BanPlayerPacket;
+import puregero.multipaper.externalserverprotocol.PardonPlayerPacket;
+
 public class UserBanList extends StoredUserList<GameProfile, UserBanListEntry> {
     public UserBanList(File file) {
         super(file);
@@ -28,4 +31,18 @@ public class UserBanList extends StoredUserList<GameProfile, UserBanListEntry> {
     protected String getKeyForUser(GameProfile gameProfile) {
         return gameProfile.getId().toString();
     }
+
+    // MultiPaper start
+    @Override
+    public void add(UserBanListEntry entry) {
+        super.add(entry);
+        BanPlayerPacket.broadcast(entry.getUser().getName(), entry.getUser().getId(), entry.created, entry.expires, entry.reason, entry.source);
+    }
+
+    @Override
+    public void remove(GameProfile key) {
+        super.remove(key);
+        PardonPlayerPacket.broadcast(key.getName(), key.getId());
+    }
+    // MultiPaper end
 }
diff --git a/src/main/java/net/minecraft/server/players/UserWhiteList.java b/src/main/java/net/minecraft/server/players/UserWhiteList.java
index aaa3d5769e0e184e19d01109a76a65be634f830d..4498297bf55012b79a67920eaa2c77064a3e8d0e 100644
--- a/src/main/java/net/minecraft/server/players/UserWhiteList.java
+++ b/src/main/java/net/minecraft/server/players/UserWhiteList.java
@@ -5,6 +5,9 @@ import com.mojang.authlib.GameProfile;
 import java.io.File;
 import java.util.Objects;
 
+import puregero.multipaper.externalserverprotocol.RemoveWhiteListedPlayerPacket;
+import puregero.multipaper.externalserverprotocol.WhiteListPlayerPacket;
+
 public class UserWhiteList extends StoredUserList<GameProfile, UserWhiteListEntry> {
     public UserWhiteList(File file) {
         super(file);
@@ -36,6 +39,7 @@ public class UserWhiteList extends StoredUserList<GameProfile, UserWhiteListEntr
         }
 
         super.add(entry);
+        WhiteListPlayerPacket.broadcast(entry.getUser().getName(), entry.getUser().getId()); // MultiPaper
     }
 
     @Override
@@ -45,6 +49,7 @@ public class UserWhiteList extends StoredUserList<GameProfile, UserWhiteListEntr
         }
 
         super.remove(profile);
+        RemoveWhiteListedPlayerPacket.broadcast(profile.getName(), profile.getId()); // MultiPaper
     }
     // Paper end - Add whitelist events
 }
diff --git a/src/main/java/puregero/multipaper/ChunkKey.java b/src/main/java/puregero/multipaper/ChunkKey.java
new file mode 100644
index 0000000000000000000000000000000000000000..d139536be7ab15293b798af115e71a8dbe4a2f3d
--- /dev/null
+++ b/src/main/java/puregero/multipaper/ChunkKey.java
@@ -0,0 +1,34 @@
+package puregero.multipaper;
+
+public class ChunkKey {
+    private final String name;
+    private final int x;
+    private final int z;
+
+    public ChunkKey(String name, int x, int z) {
+        this.name = name;
+        this.x = x;
+        this.z = z;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (other instanceof ChunkKey) {
+            return ((ChunkKey) other).name.equals(name)
+                    && ((ChunkKey) other).x == x
+                    && ((ChunkKey) other).z == z;
+        }
+
+        return super.equals(other);
+    }
+
+    @Override
+    public int hashCode() {
+        // Taken from ChunkCoordIntPair
+        int i = 1664525 * this.x + 1013904223;
+        int j = 1664525 * (this.z ^ -559038737) + 1013904223;
+
+        return name.hashCode() ^ i ^ j;
+    }
+}
+
diff --git a/src/main/java/puregero/multipaper/MultiPaperChunkHandler.java b/src/main/java/puregero/multipaper/MultiPaperChunkHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..05afe1d1b254c5ac1c10630ca6ce26fdaeffcd3a
--- /dev/null
+++ b/src/main/java/puregero/multipaper/MultiPaperChunkHandler.java
@@ -0,0 +1,251 @@
+package puregero.multipaper;
+
+import ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray;
+import ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder;
+import io.papermc.paper.util.MCUtil;
+import io.papermc.paper.world.ChunkEntitySlices;
+import net.minecraft.core.BlockPos;
+import net.minecraft.network.protocol.Packet;
+import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket;
+import net.minecraft.network.protocol.game.ClientboundBlockUpdatePacket;
+import net.minecraft.network.protocol.game.ClientboundLightUpdatePacket;
+import net.minecraft.network.protocol.game.ClientboundSectionBlocksUpdatePacket;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.LightLayer;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.chunk.ChunkAccess;
+import net.minecraft.world.level.chunk.ImposterProtoChunk;
+import net.minecraft.world.level.chunk.LevelChunk;
+import net.minecraft.world.level.lighting.LevelLightEngine;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.CraftWorld;
+import puregero.multipaper.externalserverprotocol.SendUpdatePacket;
+import puregero.multipaper.mastermessagingprotocol.messages.masterbound.SubscribeChunkMessage;
+import puregero.multipaper.mastermessagingprotocol.messages.masterbound.UnsubscribeChunkMessage;
+import puregero.multipaper.mastermessagingprotocol.messages.serverbound.BooleanMessageReply;
+
+import javax.annotation.Nullable;
+import java.util.BitSet;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.UUID;
+
+public class MultiPaperChunkHandler {
+
+    private static final Logger LOGGER = LogManager.getLogger(MultiPaperChunkHandler.class.getSimpleName());
+
+    public static boolean shouldTick(Level level, BlockPos pos) {
+        LevelChunk chunk = level.getChunkIfLoaded(pos);
+        return MultiPaper.isChunkLocal(chunk);
+    }
+
+    public static void onChunkLoad(LevelChunk chunk) {
+
+    }
+
+    public static void onChunkUnload(NewChunkHolder newChunkHolder, @Nullable ChunkAccess chunk, @Nullable ChunkEntitySlices chunkEntitySlices) {
+        if (newChunkHolder.hasExternalLockRequest) {
+            MultiPaper.unlockChunk(newChunkHolder, chunk, chunkEntitySlices);
+        }
+        MultiPaper.getConnection().sendAndAwaitReply(new UnsubscribeChunkMessage(newChunkHolder.world.getWorld().getName(), newChunkHolder.chunkX, newChunkHolder.chunkZ), BooleanMessageReply.class).thenRun(() ->
+            onChunkUnsubscribed(newChunkHolder.world.getWorld().getName(), newChunkHolder.chunkX, newChunkHolder.chunkZ)
+        );
+    }
+
+    public static void onChunkUnsubscribed(String world, int cx, int cz) {
+        NewChunkHolder holder = MultiPaper.getChunkHolder(world, cx, cz);
+
+        if (holder != null && holder.getCurrentChunk() != null) {
+            LOGGER.warn("Chunk " + world + "," + cx + "," + cz + " was unsubscribed from but has a chunk loaded! Resubscribing...");
+            MultiPaper.getConnection().send(new SubscribeChunkMessage(world, cx, cz));
+        }
+    }
+
+    private static final HashSet<BlockEntity> blockEntitiesToBroadcast = new HashSet<>();
+    public static void broadcastBlockEntityChange(BlockEntity entity) {
+        if (blockUpdateChunk != null) return; // Don't broadcast the update to other servers if we're handling an update
+        blockEntitiesToBroadcast.add(entity);
+        // Wait a bit as the block entity may get changed multiple times in 1 tick
+        MultiPaper.runSync(() -> {
+            for (BlockEntity blockEntity : blockEntitiesToBroadcast) {
+                if (blockEntity != null && blockEntity.getLevel() != null) {
+                    onBlockUpdate(MultiPaper.getChunkHolder(blockEntity.getLevel().getWorld().getName(), blockEntity.getBlockPos()), ClientboundBlockEntityDataPacket.create(blockEntity, BlockEntity::saveWithFullMetadata, false));
+                }
+            }
+            blockEntitiesToBroadcast.clear();
+        });
+    }
+
+    public static void onBlockUpdate(NewChunkHolder newChunkHolder, Packet<?> packet) {
+        if (newChunkHolder == null) {
+            // Chunk is still loading
+            return;
+        }
+
+        ChunkAccess chunk = newChunkHolder.getCurrentChunk();
+
+        if (chunk instanceof ImposterProtoChunk imposterProtoChunk) {
+            chunk = imposterProtoChunk.getWrapped();
+        }
+
+        if (chunk == null) {
+            LOGGER.warn("A " + packet.getClass().getSimpleName() + " occurred on an unloaded chunk " + newChunkHolder);
+            return;
+        }
+        if (blockUpdateChunk == null) { // Don't broadcast the update to other servers if we're handling an update
+            MultiPaper.broadcastPacketToExternalServers(newChunkHolder.externalSubscribers, () -> new SendUpdatePacket(newChunkHolder.world.uuid, packet));
+        }
+    }
+
+    public static ChunkAccess blockUpdateChunk = null;
+    private static NewChunkHolder holder = null;
+    public static void handleBlockUpdate(UUID world, Packet<?> packet, int depth) {
+        holder = null;
+        blockUpdateChunk = null;
+        ChunkAccess tempChunk = null;
+        CraftWorld bukkitWorld = ((CraftWorld) Bukkit.getWorld(world));
+        ServerLevel level = bukkitWorld != null ? bukkitWorld.getHandle() : null;
+        if (level == null) {
+            return;
+        } else if (packet instanceof ClientboundBlockUpdatePacket update) {
+            holder = MultiPaper.getChunkHolder(world, update.getPos());
+        } else if (packet instanceof ClientboundSectionBlocksUpdatePacket update) {
+            update.runUpdates((pos, state) -> {
+                if (holder == null) holder = MultiPaper.getChunkHolder(world, pos);
+            });
+        } else if (packet instanceof ClientboundBlockEntityDataPacket update) {
+            holder = MultiPaper.getChunkHolder(world, update.getPos());
+        } else if (packet instanceof ClientboundLightUpdatePacket update) {
+            holder = MultiPaper.getChunkHolder(world, update.getX(), update.getZ());
+        }
+
+        if (holder != null && holder.hasGenerationTask()) {
+            holder.onCurrentChunkLoaded().thenRun(() -> {
+                if (!Bukkit.isPrimaryThread()) {
+                    new Exception("Loaded chunk " + holder.chunkX + "," + holder.chunkZ + ", outside of the main thread! (currentThread=" + Thread.currentThread() + ")").printStackTrace();
+                }
+                handleBlockUpdate(world, packet, depth);
+            });
+            return;
+        }
+
+        if (holder != null) {
+            tempChunk = holder.getCurrentChunk();
+
+            if (tempChunk instanceof ImposterProtoChunk imposterProtoChunk) {
+                tempChunk = imposterProtoChunk.getWrapped();
+            }
+        }
+
+        if (holder != null) {
+            // Only send changes we make below to players, not other servers
+            holder.vanillaChunkHolder.addChangesToPlayersOnly = true;
+        }
+
+        // Set blockUpdateChunk here so that we can broadcast changes beforehand
+        blockUpdateChunk = tempChunk;
+
+        if (holder != null && level.getChunkIfLoaded(holder.chunkX, holder.chunkZ) != null) {
+            // Chunk is loaded
+            if (packet instanceof ClientboundBlockUpdatePacket update) {
+                setBlock(((LevelChunk) blockUpdateChunk), update.getPos(), update.getBlockState());
+            } else if (packet instanceof ClientboundSectionBlocksUpdatePacket update) {
+                update.runUpdates((pos, state) -> {
+                    setBlock(((LevelChunk) blockUpdateChunk), pos, state);
+                });
+            } else if (packet instanceof ClientboundBlockEntityDataPacket update) {
+                BlockEntity existingBlockEntity = blockUpdateChunk.getBlockEntity(update.getPos());
+                if (existingBlockEntity != null && existingBlockEntity.minecraftKey.toString().equals(update.getTag().getString("id"))) {
+                    existingBlockEntity.load(update.getTag());
+                    holder.vanillaChunkHolder.blockChanged(update.getPos());
+                } else if (!blockUpdateChunk.getBlockState(update.getPos()).hasBlockEntity() && depth < 1) {
+                    MCUtil.scheduleTask(1, () -> handleBlockUpdate(world, packet, depth + 1));
+                } else {
+                    blockUpdateChunk.removeBlockEntity(update.getPos());
+                    blockUpdateChunk.setBlockEntityNbt(update.getTag());
+                    blockUpdateChunk.getBlockEntity(update.getPos());
+                    holder.vanillaChunkHolder.blockChanged(update.getPos());
+                }
+            }
+        } else if (blockUpdateChunk != null) {
+            // Chunk is not loaded
+            if (packet instanceof ClientboundBlockUpdatePacket update) {
+                setBlockInUnloadedChunk(blockUpdateChunk, update.getPos(), update.getBlockState());
+            } else if (packet instanceof ClientboundSectionBlocksUpdatePacket update) {
+                update.runUpdates((pos, state) -> {
+                    setBlockInUnloadedChunk(blockUpdateChunk, pos, state);
+                });
+            } else if (packet instanceof ClientboundBlockEntityDataPacket update) {
+                blockUpdateChunk.removeBlockEntity(update.getPos());
+                if (blockUpdateChunk instanceof LevelChunk levelChunk && levelChunk.blockEntitiesToLoad != null) {
+                    levelChunk.blockEntitiesToLoad.add(update.getTag());
+                } else {
+                    blockUpdateChunk.setBlockEntityNbt(update.getTag());
+                    blockUpdateChunk.getBlockEntity(update.getPos());
+                }
+            } else if (packet instanceof ClientboundLightUpdatePacket update) {
+                handleLightUpdatePacket(level, blockUpdateChunk, update);
+            }
+        }
+
+        if (holder != null) {
+            holder.vanillaChunkHolder.addChangesToPlayersOnly = false;
+        }
+
+        blockUpdateChunk = null;
+    }
+
+    private static void setBlockInUnloadedChunk(ChunkAccess chunk, BlockPos pos, BlockState blockState) {
+        if (chunk instanceof LevelChunk levelChunk) {
+            levelChunk.setBlockState(pos, blockState, false, false);
+        } else {
+            chunk.setBlockState(pos, blockState, false);
+        }
+    }
+
+    private static void setBlock(LevelChunk chunk, BlockPos pos, BlockState blockState) {
+        BlockState oldState = chunk.setBlockState(pos, blockState, false, false);
+        holder.vanillaChunkHolder.blockChanged(pos);
+
+        if (oldState != null && blockState != oldState && (blockState.getLightBlock(chunk, pos) != oldState.getLightBlock(chunk, pos) || blockState.getLightEmission() != oldState.getLightEmission() || blockState.useShapeForLightOcclusion() || oldState.useShapeForLightOcclusion())) {
+            chunk.level.getProfiler().push("queueCheckLightExternalUpdate");
+            chunk.level.getChunkSource().getLightEngine().checkBlock(pos);
+            chunk.level.getProfiler().pop();
+        }
+    }
+
+    // From the client
+    private static void handleLightUpdatePacket(ServerLevel level, ChunkAccess chunk, ClientboundLightUpdatePacket packet) {
+        int i = packet.getX();
+        int j = packet.getZ();
+        LevelLightEngine levellightengine = level.getChunkSource().getLightEngine();
+        BitSet bitset = packet.getLightData().getSkyYMask();
+        BitSet bitset1 = packet.getLightData().getEmptySkyYMask();
+        Iterator<byte[]> iterator = packet.getLightData().getSkyUpdates().iterator();
+        readSectionList(chunk, i, j, levellightengine, LightLayer.SKY, bitset, bitset1, iterator);
+        BitSet bitset2 = packet.getLightData().getBlockYMask();
+        BitSet bitset3 = packet.getLightData().getEmptyBlockYMask();
+        Iterator<byte[]> iterator1 = packet.getLightData().getBlockUpdates().iterator();
+        readSectionList(chunk, i, j, levellightengine, LightLayer.BLOCK, bitset2, bitset3, iterator1);
+    }
+
+    // From the client
+    private static void readSectionList(ChunkAccess chunk, int i, int j, LevelLightEngine levelLightEngine, LightLayer lightLayer, BitSet bitset2, BitSet bitset3, Iterator<byte[]> iterator1) {
+        for(int k = 0; k < levelLightEngine.getLightSectionCount(); ++k) {
+            int l = levelLightEngine.getMinLightSection() + k;
+            boolean flag = bitset2.get(k);
+            boolean flag1 = bitset3.get(k);
+            if (flag || flag1) {
+                if (lightLayer == LightLayer.BLOCK) {
+                    chunk.getBlockNibbles()[k] = flag ? new SWMRNibbleArray(iterator1.next().clone()) : new SWMRNibbleArray();
+                } else if (lightLayer == LightLayer.SKY) {
+                    chunk.getSkyNibbles()[k] = flag ? new SWMRNibbleArray(iterator1.next().clone()) : new SWMRNibbleArray();
+                }
+            }
+        }
+    }
+}
diff --git a/src/main/java/puregero/multipaper/MultiPaperWorldBorderHandler.java b/src/main/java/puregero/multipaper/MultiPaperWorldBorderHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..229506e75000777a5c3c686533eac47ef17b30d4
--- /dev/null
+++ b/src/main/java/puregero/multipaper/MultiPaperWorldBorderHandler.java
@@ -0,0 +1,78 @@
+package puregero.multipaper;
+
+import net.minecraft.network.protocol.Packet;
+import net.minecraft.network.protocol.game.*;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.level.border.BorderChangeListener;
+import net.minecraft.world.level.border.WorldBorder;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.CraftWorld;
+import org.jetbrains.annotations.NotNull;
+import puregero.multipaper.externalserverprotocol.SendUpdatePacket;
+
+import java.util.UUID;
+
+public class MultiPaperWorldBorderHandler implements BorderChangeListener {
+
+    public static boolean updatingWorldBorder = false;
+
+    @Override
+    public void onBorderSizeSet(@NotNull WorldBorder border, double size) {
+        onWorldBorderChange(border, new ClientboundSetBorderSizePacket(border));
+    }
+
+    @Override
+    public void onBorderSizeLerping(@NotNull WorldBorder border, double fromSize, double toSize, long time) {
+        onWorldBorderChange(border, new ClientboundSetBorderLerpSizePacket(border));
+    }
+
+    @Override
+    public void onBorderCenterSet(@NotNull WorldBorder border, double centerX, double centerZ) {
+        onWorldBorderChange(border, new ClientboundSetBorderCenterPacket(border));
+    }
+
+    @Override
+    public void onBorderSetWarningTime(@NotNull WorldBorder border, int warningTime) {
+        onWorldBorderChange(border, new ClientboundSetBorderWarningDelayPacket(border));
+    }
+
+    @Override
+    public void onBorderSetWarningBlocks(@NotNull WorldBorder border, int warningBlockDistance) {
+        onWorldBorderChange(border, new ClientboundSetBorderWarningDistancePacket(border));
+    }
+
+    @Override
+    public void onBorderSetDamagePerBlock(@NotNull WorldBorder border, double damagePerBlock) {}
+
+    @Override
+    public void onBorderSetDamageSafeZOne(@NotNull WorldBorder border, double safeZoneRadius) {}
+
+    private void onWorldBorderChange(WorldBorder border, Packet<?> packet) {
+        if (!updatingWorldBorder && border.world != null && border.world.getWorldBorder() == border) {
+            MultiPaper.broadcastPacketToExternalServers(border.world.getWorld().getName(), new SendUpdatePacket(border.world.uuid, packet));
+
+            // Save the level.dat
+            border.world.saveIncrementally(true);
+        }
+    }
+
+    public static void handle(UUID world, Packet<?> packet) {
+        ServerLevel level = ((CraftWorld) Bukkit.getWorld(world)).getHandle();
+
+        updatingWorldBorder = true;
+
+        if (packet instanceof ClientboundSetBorderSizePacket setBorderSizePacket) {
+            level.getWorldBorder().setSize(setBorderSizePacket.getSize());
+        } else if (packet instanceof ClientboundSetBorderLerpSizePacket setBorderLerpSizePacket) {
+            level.getWorldBorder().lerpSizeBetween(setBorderLerpSizePacket.getOldSize(), setBorderLerpSizePacket.getNewSize(), setBorderLerpSizePacket.getLerpTime());
+        } else if (packet instanceof ClientboundSetBorderCenterPacket setBorderCenterPacket) {
+            level.getWorldBorder().setCenter(setBorderCenterPacket.getNewCenterX(), setBorderCenterPacket.getNewCenterZ());
+        } else if (packet instanceof ClientboundSetBorderWarningDelayPacket setBorderWarningDelayPacket) {
+            level.getWorldBorder().setWarningTime(setBorderWarningDelayPacket.getWarningDelay());
+        } else if (packet instanceof ClientboundSetBorderWarningDistancePacket setBorderWarningDistancePacket) {
+            level.getWorldBorder().setWarningBlocks(setBorderWarningDistancePacket.getWarningBlocks());
+        }
+
+        updatingWorldBorder = false;
+    }
+}
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index 361232ab565973d262cfb6ecc54fd823e28d799e..773d03d3500b868fd6d0ede5a8ccc63885fa4a04 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -20,6 +20,18 @@ public class ExternalServerPacketSerializer {
         addPacket(PlayerRespawnPacket.class, PlayerRespawnPacket::new);
         addPacket(PlayerChangeDimensionPacket.class, PlayerChangeDimensionPacket::new);
         addPacket(SendPacketPacket.class, SendPacketPacket::new);
+        addPacket(RequestChunkPacket.class, RequestChunkPacket::new);
+        addPacket(SendChunkPacket.class, SendChunkPacket::new);
+        addPacket(SendTickListPacket.class, SendTickListPacket::new);
+        addPacket(SendUpdatePacket.class, SendUpdatePacket::new);
+        addPacket(BanPlayerPacket.class, BanPlayerPacket::new);
+        addPacket(PardonPlayerPacket.class, PardonPlayerPacket::new);
+        addPacket(BanIpPacket.class, BanIpPacket::new);
+        addPacket(PardonIpPacket.class, PardonIpPacket::new);
+        addPacket(OpPlayerPacket.class, OpPlayerPacket::new);
+        addPacket(DeopPlayerPacket.class, DeopPlayerPacket::new);
+        addPacket(WhiteListPlayerPacket.class, WhiteListPlayerPacket::new);
+        addPacket(RemoveWhiteListedPlayerPacket.class, RemoveWhiteListedPlayerPacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/RequestChunkPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/RequestChunkPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..4bfe8626ccfc5de2918a3b3416c884e74b26cba3
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/RequestChunkPacket.java
@@ -0,0 +1,146 @@
+package puregero.multipaper.externalserverprotocol;
+
+import io.papermc.paper.chunk.system.scheduling.NewChunkHolder;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.ListTag;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ThreadedLevelLightEngine;
+import net.minecraft.world.level.chunk.ChunkAccess;
+import net.minecraft.world.level.chunk.ImposterProtoChunk;
+import net.minecraft.world.level.chunk.LevelChunk;
+import net.minecraft.world.level.chunk.storage.ChunkSerializer;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.World;
+import org.bukkit.craftbukkit.CraftWorld;
+import puregero.multipaper.ExternalServer;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.util.ConcurrentModificationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class RequestChunkPacket extends ExternalServerPacket {
+
+    private static final Logger LOGGER = LogManager.getLogger(RequestChunkPacket.class.getSimpleName());
+    public static ExternalServer blocker = null;
+
+    private final String world;
+    private final int cx;
+    private final int cz;
+
+    public RequestChunkPacket(String world, int cx, int cz) {
+        this.world = world;
+        this.cx = cx;
+        this.cz = cz;
+    }
+
+    public RequestChunkPacket(FriendlyByteBuf in) {
+        world = in.readUtf();
+        cx = in.readInt();
+        cz = in.readInt();
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUtf(world);
+        out.writeInt(cx);
+        out.writeInt(cz);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        World bukkitWorld = Bukkit.getWorld(world);
+
+        if (!(bukkitWorld instanceof CraftWorld craftWorld)) {
+            LOGGER.warn(connection.externalServer.getName() + " is requesting chunk " + world + "," + cx + "," + cz + " but we don't have the world " + world + " loaded.");
+            connection.send(new SendChunkPacket(world, cx, cz, null));
+            return;
+        }
+
+        ServerLevel level = craftWorld.getHandle();
+        NewChunkHolder newChunkHolder = MultiPaper.getChunkHolder(world, cx, cz);
+
+        if (newChunkHolder == null) {
+            LOGGER.warn(connection.externalServer.getName() + " is requesting chunk " + world + "," + cx + "," + cz + " but we aren't trying to load it.");
+            connection.send(new SendChunkPacket(world, cx, cz, null));
+            return;
+        }
+
+        CompletableFuture<ChunkAccess> future = newChunkHolder.onCurrentChunkLoaded();
+
+        CompletableFuture<ChunkAccess> futureToWaitOn = future;
+        if (blocker == connection.externalServer) {
+            ChunkAccess access = newChunkHolder.getCurrentChunk();
+            if (access != null) {
+                futureToWaitOn = CompletableFuture.completedFuture(newChunkHolder.getCurrentChunk());
+            } else {
+                connection.send(new SendChunkPacket(world, cx, cz, null));
+                return;
+            }
+        }
+
+        futureToWaitOn.thenAccept(chunk -> {
+            if (future != newChunkHolder.onCurrentChunkLoaded()) {
+                // The future has been updated, try again
+                handle(connection);
+                return;
+            }
+
+            if (chunk == null) {
+                LOGGER.warn(connection.externalServer.getName() + " is requesting chunk " + world + "," + cx + "," + cz + " but we don't have it loaded.");
+                connection.send(new SendChunkPacket(world, cx, cz, null));
+                return;
+            }
+
+            CompletableFuture<Void> lightFuture = ((ThreadedLevelLightEngine) level.getLightEngine()).theLightEngine.lightQueue.getChunkFuture(newChunkHolder.vanillaChunkHolder.getPos());
+
+            if (!lightFuture.isDone()) {
+                lightFuture.thenRunAsync(() -> handle(connection));
+                return;
+            }
+
+            try {
+                ListTag entitiesToLoad = null;
+                ListTag blockEntitiesToLoad = null;
+
+                ChunkAccess fullChunk = chunk instanceof ImposterProtoChunk imposterProtoChunk ? imposterProtoChunk.getWrapped() : chunk;
+
+                if (fullChunk instanceof LevelChunk levelChunk) {
+                    // Cache these tags in case they get deleted while we serialize the chunk (multi-threaded fun!)
+                    entitiesToLoad = levelChunk.entitiesToLoad;
+                    blockEntitiesToLoad = levelChunk.blockEntitiesToLoad;
+                }
+
+                CompoundTag tag = ChunkSerializer.write(level, chunk);
+
+                if (entitiesToLoad != null) {
+                    tag.put("entities", entitiesToLoad);
+                }
+                if (blockEntitiesToLoad != null) {
+                    tag.put("block_entities", blockEntitiesToLoad);
+                }
+
+                connection.send(new SendChunkPacket(world, cx, cz, tag));
+            } catch (ConcurrentModificationException e) {
+                LOGGER.warn("Got ConcurrentModificationException while sending chunk, sending it in main thread instead");
+                MultiPaper.runSync(() -> handle(connection));
+            }
+        })
+        // Timeout instantly if this server is blocking our chunk loading, as this is probably also blocking their chunk loading
+        .orTimeout(15, TimeUnit.SECONDS).exceptionally(throwable -> {
+            if (throwable instanceof TimeoutException) {
+                LOGGER.warn("Timed out while sending chunk " + world + "," + cx + "," + cz);
+            } else {
+                LOGGER.warn("Error while sending chunk " + world + "," + cx + "," + cz, throwable);
+            }
+
+            connection.send(new SendChunkPacket(world, cx, cz, null));
+            return null;
+        });
+    }
+}
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/SendChunkPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/SendChunkPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..8ed4aebd2a08e9dbcb1e5d17fe30eadb87428a74
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/SendChunkPacket.java
@@ -0,0 +1,122 @@
+package puregero.multipaper.externalserverprotocol;
+
+import io.papermc.paper.chunk.system.scheduling.NewChunkHolder;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.level.ChunkHolder;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.chunk.ChunkAccess;
+import net.minecraft.world.level.chunk.ImposterProtoChunk;
+import net.minecraft.world.level.chunk.LevelChunk;
+import net.minecraft.world.level.chunk.storage.ChunkSerializer;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.CraftWorld;
+import puregero.multipaper.ChunkKey;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.util.function.Consumer;
+import java.util.zip.InflaterInputStream;
+
+public class SendChunkPacket extends ExternalServerPacket {
+
+    private static final Logger LOGGER = LogManager.getLogger(SendChunkPacket.class.getSimpleName());
+
+    private final String world;
+    private final int cx;
+    private final int cz;
+    private final byte[] data;
+
+    public SendChunkPacket(String world, int cx, int cz, CompoundTag tag) {
+        this.world = world;
+        this.cx = cx;
+        this.cz = cz;
+
+        try {
+            if (tag == null) {
+                data = new byte[0];
+            } else {
+                data = MultiPaper.nbtToBytes(tag);
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public SendChunkPacket(FriendlyByteBuf in) {
+        world = in.readUtf();
+        cx = in.readInt();
+        cz = in.readInt();
+        data = in.readByteArray();
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUtf(world);
+        out.writeInt(cx);
+        out.writeInt(cz);
+        out.writeByteArray(data);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        try {
+            Consumer<DataInputStream> callback = connection.chunkCallbacks.remove(new ChunkKey(world, cx, cz));
+            if (callback != null) {
+                if (data.length == 0) {
+                    LOGGER.warn(connection.externalServer.getName() + " sent us an empty chunk for " + world + "," + cx + "," + cz + ", force loading it from disk");
+                    MultiPaper.forceReadChunk(world, "region", cx, cz).thenAccept(data2 -> callback.accept(data2.length == 0 ? null : new DataInputStream(new InflaterInputStream(new ByteArrayInputStream(data2)))));
+                } else {
+                    callback.accept(new DataInputStream(new ByteArrayInputStream(data)));
+                }
+            } else {
+                if (data.length == 0) {
+                    return;
+                }
+
+                CompoundTag tag = MultiPaper.nbtFromBytes(data);
+                CraftWorld craftWorld = ((CraftWorld) Bukkit.getWorld(world));
+                ServerLevel level = craftWorld != null ? craftWorld.getHandle() : null;
+                NewChunkHolder holder = MultiPaper.getChunkHolder(world, cx, cz);
+                if (level == null) {
+                    LOGGER.warn("Received chunk data " + world + "," + cx + "," + cz + " but we don't have world loaded.");
+                } else if (holder == null) {
+                    LOGGER.warn("Received chunk data " + world + "," + cx + "," + cz + " but no chunk is loaded here");
+                } else if (holder.getCurrentChunk() instanceof LevelChunk) {
+                    LOGGER.warn("Received chunk data " + world + "," + cx + "," + cz + " (" + tag.getString("Status") + "), but it is a level chunk (" + holder.getCurrentGenStatus() + ")");
+                } else {
+                    LOGGER.warn("Received chunk data " + world + "," + cx + "," + cz + " (" + tag.getString("Status") + "), but we have a " + holder.getCurrentGenStatus() + " chunk, forcing reload from disk.");
+                    forceChunkUnsafeUnload(world, cx, cz);
+                }
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private ChunkAccess loadChunk(ServerLevel level, ChunkPos pos, CompoundTag tag) {
+        ChunkSerializer.InProgressChunkHolder newChunk = ChunkSerializer.loadChunk(level, level.getPoiManager(), pos, tag, true);
+        return newChunk.protoChunk instanceof ImposterProtoChunk imposterProtoChunk ? imposterProtoChunk.getWrapped() : newChunk.protoChunk;
+    }
+
+    public static void forceChunkUnsafeUnload(String world, int cx, int cz) {
+        MultiPaper.runSync(() -> {
+            ChunkPos pos = new ChunkPos(cx, cz);
+            CraftWorld craftWorld = ((CraftWorld) Bukkit.getWorld(world));
+            if (craftWorld != null) {
+                ServerLevel level = craftWorld.getHandle();
+                NewChunkHolder holder = level.chunkTaskScheduler.chunkHolderManager.getChunkHolder(pos.longKey);
+
+                if (holder != null) {
+                    level.chunkTaskScheduler.chunkHolderManager.unloadChunkNowNoSave(holder);
+                }
+            }
+        });
+    }
+}
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/SendTickListPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/SendTickListPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..801d03669f9056ca9cd3dc2c617521c29a6854be
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/SendTickListPacket.java
@@ -0,0 +1,89 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.Tag;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.level.chunk.ChunkAccess;
+import net.minecraft.world.level.chunk.LevelChunk;
+import net.minecraft.world.level.chunk.ProtoChunk;
+import net.minecraft.world.ticks.LevelChunkTicks;
+import net.minecraft.world.ticks.ProtoChunkTicks;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.CraftWorld;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.io.IOException;
+
+public class SendTickListPacket extends ExternalServerPacket {
+
+    private static final Logger LOGGER = LogManager.getLogger(SendTickListPacket.class.getSimpleName());
+
+    private final String world;
+    private final int cx;
+    private final int cz;
+    private final CompoundTag tag;
+
+    public SendTickListPacket(LevelChunk chunk) {
+        this.world = chunk.level.convertable.getLevelId();
+        this.cx = chunk.locX;
+        this.cz = chunk.locZ;
+
+        tag = new CompoundTag();
+        tag.put("block_ticks", chunk.blockTicks.save(chunk.level.getLevelData().getGameTime(), (block) -> BuiltInRegistries.BLOCK.getKey(block).toString()));
+        tag.put("fluid_ticks", chunk.fluidTicks.save(chunk.level.getLevelData().getGameTime(), (fluidtype) -> BuiltInRegistries.FLUID.getKey(fluidtype).toString()));
+    }
+
+    public SendTickListPacket(FriendlyByteBuf in) {
+        this.world = in.readUtf();
+        this.cx = in.readInt();
+        this.cz = in.readInt();
+
+        byte[] data = in.readByteArray();
+        try {
+            tag = MultiPaper.nbtFromBytes(data);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUtf(world);
+        out.writeInt(cx);
+        out.writeInt(cz);
+
+        try {
+            byte[] data = MultiPaper.nbtToBytes(tag);
+
+            out.writeByteArray(data);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            ServerLevel level = ((CraftWorld) Bukkit.getWorld(world)).getHandle();
+            ChunkAccess chunk = MultiPaper.getChunkAccess(world, cx, cz);
+            if (chunk instanceof LevelChunk levelChunk) {
+                levelChunk.unregisterTickContainerFromLevel(level);
+                levelChunk.blockTicks = LevelChunkTicks.load(tag.getList("block_ticks", Tag.TAG_COMPOUND), s -> BuiltInRegistries.BLOCK.getOptional(ResourceLocation.tryParse(s)), chunk.getPos());
+                levelChunk.fluidTicks = LevelChunkTicks.load(tag.getList("fluid_ticks", Tag.TAG_COMPOUND), s -> BuiltInRegistries.FLUID.getOptional(ResourceLocation.tryParse(s)), chunk.getPos());
+                levelChunk.unpackTicks(level.getLevelData().getGameTime());
+                if (levelChunk.loaded) levelChunk.registerTickContainerInLevel(level);
+            } else if (chunk instanceof ProtoChunk protoChunk) {
+                protoChunk.blockTicks = ProtoChunkTicks.load(tag.getList("block_ticks", Tag.TAG_COMPOUND), s -> BuiltInRegistries.BLOCK.getOptional(ResourceLocation.tryParse(s)), chunk.getPos());
+                protoChunk.fluidTicks = ProtoChunkTicks.load(tag.getList("fluid_ticks", Tag.TAG_COMPOUND), s -> BuiltInRegistries.FLUID.getOptional(ResourceLocation.tryParse(s)), chunk.getPos());
+            } else {
+                LOGGER.warn("Received tick lists for an unloaded chunk " + world + "," + cx + "," + cz);
+            }
+        });
+    }
+}
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/SendUpdatePacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/SendUpdatePacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..c9ad59beb1115af9b058b3a03e41705f74df2cba
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/SendUpdatePacket.java
@@ -0,0 +1,65 @@
+package puregero.multipaper.externalserverprotocol;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import net.minecraft.network.ConnectionProtocol;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.network.protocol.Packet;
+import net.minecraft.network.protocol.PacketFlow;
+import net.minecraft.network.protocol.game.*;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+import puregero.multipaper.MultiPaperChunkHandler;
+import puregero.multipaper.MultiPaperWorldBorderHandler;
+
+import java.util.UUID;
+
+public class SendUpdatePacket extends ExternalServerPacket {
+
+    private static final Logger LOGGER = LogManager.getLogger(SendUpdatePacket.class.getSimpleName());
+
+    private final UUID world;
+    private final Packet<?> packet;
+
+    public SendUpdatePacket(UUID world, Packet<?> packet) {
+        this.world = world;
+        this.packet = packet;
+    }
+
+    public SendUpdatePacket(FriendlyByteBuf in) {
+        world = in.readUUID();
+
+        byte[] bytes = in.readByteArray();
+        ByteBuf buf = Unpooled.wrappedBuffer(bytes);
+        FriendlyByteBuf friendlyByteBuf = new FriendlyByteBuf(buf);
+        friendlyByteBuf.maxNbtSize = Long.MAX_VALUE;
+        int packetId = friendlyByteBuf.readVarInt();
+        packet = ConnectionProtocol.PLAY.createPacket(PacketFlow.CLIENTBOUND, packetId, friendlyByteBuf);
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUUID(world);
+
+        ConnectionProtocol protocol = ConnectionProtocol.getProtocolForPacket(packet);
+        Integer id = protocol.getPacketId(PacketFlow.CLIENTBOUND, packet);
+        ByteBuf buf = Unpooled.buffer();
+        FriendlyByteBuf friendlyByteBuf = new FriendlyByteBuf(buf);
+        friendlyByteBuf.writeVarInt(id);
+        packet.write(friendlyByteBuf);
+        out.writeByteArray(buf.array());
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        if (packet instanceof ClientboundBlockUpdatePacket || packet instanceof ClientboundSectionBlocksUpdatePacket || packet instanceof ClientboundLightUpdatePacket || packet instanceof ClientboundBlockEntityDataPacket) {
+            MultiPaper.runSync(() -> MultiPaperChunkHandler.handleBlockUpdate(world, packet, 0));
+        } else if (packet instanceof ClientboundSetBorderSizePacket || packet instanceof ClientboundSetBorderLerpSizePacket || packet instanceof ClientboundSetBorderCenterPacket || packet instanceof ClientboundSetBorderWarningDelayPacket || packet instanceof ClientboundSetBorderWarningDistancePacket) {
+            MultiPaper.runSync(() -> MultiPaperWorldBorderHandler.handle(world, packet));
+        } else {
+            LOGGER.warn("Unhandled update packet of type " + packet.getClass().getSimpleName());
+        }
+    }
+}
